File: Sources/LanguageServer/FoldingRangeVisitor.swift
import Foundation
import LanguageServerProtocol
import Logging
import MesonAST

public class FoldingRangeVisitor: CodeVisitor {
  static let LOG = Logger(label: "LanguageServer::FoldingRangeVisitor")
  internal var ranges: [FoldingRange] = []

  public func visitSourceFile(file: SourceFile) { file.visitChildren(visitor: self) }

  public func visitBuildDefinition(node: BuildDefinition) { node.visitChildren(visitor: self) }

  public func visitErrorNode(node: ErrorNode) { node.visitChildren(visitor: self) }

  public func visitSelectionStatement(node: SelectionStatement) {
    node.visitChildren(visitor: self)
  }

  public func visitBreakStatement(node: BreakNode) { node.visitChildren(visitor: self) }

  public func visitContinueStatement(node: ContinueNode) { node.visitChildren(visitor: self) }

  public func visitIterationStatement(node: IterationStatement) {
    node.visitChildren(visitor: self)
    if !node.block.isEmpty {
      self.ranges.append(
        FoldingRange(
          startLine: Int(node.location.startLine),
          endLine: Int(node.location.endLine - 1)
        )
      )
    }
  }

  public func visitAssignmentStatement(node: AssignmentStatement) {
    node.visitChildren(visitor: self)
  }

  public func visitFunctionExpression(node: FunctionExpression) {
    node.visitChildren(visitor: self)
  }

  public func visitArgumentList(node: ArgumentList) { node.visitChildren(visitor: self) }

  public func visitKeywordItem(node: KeywordItem) { node.visitChildren(visitor: self) }

  public func visitConditionalExpression(node: ConditionalExpression) {
    node.visitChildren(visitor: self)
  }

  public func visitUnaryExpression(node: UnaryExpression) { node.visitChildren(visitor: self) }

  public func visitSubscriptExpression(node: SubscriptExpression) {
    node.visitChildren(visitor: self)
  }

  public func visitMethodExpression(node: MethodExpression) { node.visitChildren(visitor: self) }

  public func visitIdExpression(node: IdExpression) { node.visitChildren(visitor: self) }

  public func visitBinaryExpression(node: BinaryExpression) { node.visitChildren(visitor: self) }

  public func visitStringLiteral(node: StringLiteral) { node.visitChildren(visitor: self) }

  public func visitArrayLiteral(node: ArrayLiteral) { node.visitChildren(visitor: self) }

  public func visitBooleanLiteral(node: BooleanLiteral) { node.visitChildren(visitor: self) }

  public func visitIntegerLiteral(node: IntegerLiteral) { node.visitChildren(visitor: self) }

  public func visitDictionaryLiteral(node: DictionaryLiteral) { node.visitChildren(visitor: self) }

  public func visitKeyValueItem(node: KeyValueItem) { node.visitChildren(visitor: self) }
}

File: src/liblangserver/foldingrangevisitor.cpp
#include "foldingrangevisitor.hpp"

#include "node.hpp"

#include <cstddef>

void FoldingRangeVisitor::visitArgumentList(ArgumentList *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitArrayLiteral(ArrayLiteral *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitAssignmentStatement(AssignmentStatement *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitBinaryExpression(BinaryExpression *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitBooleanLiteral(BooleanLiteral *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitBuildDefinition(BuildDefinition *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitConditionalExpression(
    ConditionalExpression *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitDictionaryLiteral(DictionaryLiteral *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitFunctionExpression(FunctionExpression *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitIdExpression(IdExpression *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitIntegerLiteral(IntegerLiteral *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitIterationStatement(IterationStatement *node) {
  node->visitChildren(this);
  if (node->stmts.empty()) {
    return;
  }
  this->ranges.emplace_back(node->location.startLine,
                            node->location.endLine - 1);
}

void FoldingRangeVisitor::visitKeyValueItem(KeyValueItem *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitKeywordItem(KeywordItem *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitMethodExpression(MethodExpression *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitSelectionStatement(SelectionStatement *node) {
  node->visitChildren(this);

  for (size_t idx = 0; idx < node->blocks.size();) {
    auto block = node->blocks[idx];
    idx++;
    if (block.empty()) {
      continue;
    }
    auto startLine = block[0]->location.startLine - 1;
    if (idx - 1 < node->conditions.size()) {
      startLine = node->conditions[idx - 1]->location.endLine;
    }
    this->ranges.emplace_back(startLine, block.back()->location.endLine);
  }
}

void FoldingRangeVisitor::visitStringLiteral(StringLiteral *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitSubscriptExpression(SubscriptExpression *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitUnaryExpression(UnaryExpression *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitErrorNode(ErrorNode *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitBreakNode(BreakNode *node) {
  node->visitChildren(this);
}

void FoldingRangeVisitor::visitContinueNode(ContinueNode *node) {
  node->visitChildren(this);
}
